### PR TITLE
 YD-557 Pushed first and last name down

### PR DESCRIPTION
**Pull request by Bert-R**
Introduced a new message type ``BuddyConnectionChangeMessage`` and moved the first and last name there.

Along with this, provided default for missing first or last name. Messages that are created before the first and last name were moved to the private data will not have the first and last name in the messages. To ensure such messages still contain at least something reasonable, a default value is returned.

### CODE REVIEW

File: core/src/main/java/nu/yona/server/exceptions/ResourceBasedException.java [hunk 1]
```diff
@@ -100,11 +80,6 @@ protected ResourceBasedException(HttpStatus statusCode, Throwable t, String mess
 		this.statusCode = statusCode;
 	}
 
-	public static void setTranslator(Translator translator)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/exceptions/ResourceBasedException.java]: Actually, this public method is a better injection than having an inline `Translator.getInstance()`. For unit testing, it is otherwise impossible to change the dependency.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	public void decrypt()
+	{
+		super.decrypt();
+		firstName = SecretKeyUtil.decryptString(firstNameCiphertext);
+		lastName = SecretKeyUtil.decryptString(lastNameCiphertext);
+	}
+
+	public String getFirstName()
+	{
+		return (firstName == null) ? getPlaceHolderString("message.alternative.first.name") : firstName;
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java]: `getPlaceHolderString` -> `getPlaceholderString` (see [here](https://www.merriam-webster.com/dictionary/placeholder))
But I would prefer `getTranslation("message.alternative.first.name", getSenderNickname())` because it now looks like a string with placeholders is returned.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyDto.java [hunk 3]
```diff
@@ -104,16 +104,16 @@ public UserDto getUser()
 		return user;
 	}
 
-	Buddy createBuddyEntity(Translator translator)
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/BuddyDto.java]: I prefer the old method for testability. Same for `determineTempNickname`.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.Transient;
+
+import nu.yona.server.Translator;
+import nu.yona.server.crypto.seckey.SecretKeyUtil;
+import nu.yona.server.messaging.entities.BuddyMessage;
+
+@Entity
+public abstract class BuddyConnectionChangeMessage extends BuddyMessage
```
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java]: Can we add a unit test that a `BuddyConnectionChangeMessage` works as expected when `firstName` and `lastName` are `null`? So for instance:
- `getFirstName_firstNameNull_returnsAlternativeContainingNickname`
- `encrypt_firstAndLastNameNull_doesNotThrow`
- `decrypt_firstAndLastNameNull_doesNotThrow`

File: core/src/main/resources/messages/messages_nl.properties [hunk 5]
```diff
@@ -18,6 +18,8 @@ message.buddy.device.vpn.disconnect=De gebruiker heeft de VPN verbinding verbrok
 message.temp.nickname={0} {1}
 message.self.nickname={0} (ik)
 message.system.nickname=Yona
+message.alternative.first.name=<{}>
+message.alternative.last.name=<{}>
```
Comment (aristotelos) [inline core/src/main/resources/messages/messages_nl.properties]: Why not `<{0}>` as in English?

File: core/src/main/resources/messages/messages.properties [hunk 6]
```diff
@@ -18,6 +18,8 @@ message.buddy.device.vpn.disconnect=User disconnected VPN on device ''{0}''
 message.temp.nickname={0} {1}
 message.self.nickname={0} (me)
 message.system.nickname=Yona
+message.alternative.first.name=<{0}>
```
Comment (aristotelos) [inline core/src/main/resources/messages/messages.properties]: Does this show `<Nickname>` literally? I hope the app did their html escaping right...

File: core/src/main/java/nu/yona/server/exceptions/ResourceBasedException.java (hunk 1 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/exceptions/ResourceBasedException.java]: We currently don't have tests that depend on this, but if we want to do it, we can use a similar setter in ``Translator``. The code to get the instance in a static context needs to be in ``Translator``, as that's the only one to logically depend upon.

File: core/src/main/java/nu/yona/server/subscriptions/service/BuddyDto.java (hunk 3 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/service/BuddyDto.java]: For testability, we can easily set a mock instance on the static member. That way, we can keep the code clean.

File: core/src/main/resources/messages/messages.properties (hunk 6 above)
Comment (Bert-R) [reply core/src/main/resources/messages/messages.properties]: Indeed, it shows something like ``<BR>``. The app doesn't do HTML, so that's not an issue.

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java (hunk 2 above)
Comment (aristotelos) [reply core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java]: I still would prefer `getTranslation("message.alternative.first.name", getSenderNickname())` and `getTranslation("message.alternative.last.name", getSenderNickname())`.

Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java]: OK

File: core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java (hunk 4 above)
Comment (Bert-R) [reply core/src/main/java/nu/yona/server/subscriptions/entities/BuddyConnectionChangeMessage.java]: Created [YD-558](https://jira.yona.nu/browse/YD-558).

**Comment by Bert-R** [comment]
Pushed an update
